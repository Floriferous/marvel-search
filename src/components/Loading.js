import React from 'react';

const Loading = () => <p>Loading super-heroes...</p>;

export default Loading;
