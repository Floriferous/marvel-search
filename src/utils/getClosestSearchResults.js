import { createSearchKey } from '../actions/search';

/**
 * getClosestSearchResults - Searches for the closest search results
 * currently in memory. It iterates on a search string, and removes one
 * character at a time until it finds an old search result to display.
 *
 * If the first run passes, it simply returns the exact search results.
 *
 * An isLoading flag is also returned, to indicate whether old or exact
 * results are currently being returned
 *
 * @param {Object} searchResults Object with all previous search results
 * @param {String} search        current search string
 * @param {String} pagination    current pagination
 *
 * @return {Object}
 */
const getClosestSearchResults = (searchResults, search, pagination) => {
  let broaderResults;
  let isExactSearch;

  // Return immediately if search length equals 1 or pagination is larger than 0
  if (search.length === 1 || pagination > 0) {
    return {
      data: searchResults[createSearchKey(search, pagination)] || {},
      isLoading: false, // taken care of by CharacterList
    };
  }

  for (let i = 0; i < search.length; i += 1) {
    const broaderSearch = search.slice(0, search.length - i);
    broaderResults = searchResults[createSearchKey(broaderSearch, pagination)];
    if (broaderResults) {
      isExactSearch = i === 0;
      break;
    }
  }
  return { data: broaderResults || {}, isLoading: !isExactSearch };
};

export default getClosestSearchResults;
