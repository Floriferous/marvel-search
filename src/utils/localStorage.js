export const storageAvailable = () => {
  try {
    const storage = localStorage;
    const x = '__storage_test__';
    storage.setItem(x, x);
    storage.removeItem(x);
    return true;
  } catch (e) {
    return false;
  }
};

export const saveValueAtKey = (key, value) => {
  if (storageAvailable()) {
    localStorage.setItem(key, JSON.stringify(value));
    return true;
  }

  return false;
};

export const getValueAtKey = (key) => {
  if (storageAvailable()) {
    return JSON.parse(localStorage.getItem(key));
  }

  return false;
};
