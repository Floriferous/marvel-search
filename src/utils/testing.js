import React from 'react';
import Enzyme from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

Enzyme.configure({ adapter: new Adapter() });

export const testComponent = (Component, props) => {
  if (!testComponent.mountedLockScreen) {
    testComponent.mountedLockScreen = Enzyme.shallow(<Component {...props} />);
  }
  return testComponent.mountedLockScreen;
};

testComponent.reset = () => {
  testComponent.mountedLockScreen = undefined;
};

export const { shallow, mount, render } = Enzyme;
export default Enzyme;
