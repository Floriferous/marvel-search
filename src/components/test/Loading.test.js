import { testComponent } from '../../utils/testing';
import Loading from '../Loading';

describe('Loading', () => {
  let props;
  const component = () => testComponent(Loading, props);

  beforeEach(() => {
    props = {};
    testComponent.reset();
  });

  it('matches its snapshot', () => {
    expect(component()).toMatchSnapshot();
  });

  it('renders a p with some text', () => {
    expect(component().find('p').length).toBe(1);
    expect(component()
      .find('p')
      .text().length).toBeGreaterThan(0);
  });
});
