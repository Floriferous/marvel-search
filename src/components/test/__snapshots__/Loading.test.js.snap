// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Disclaimer matches its snapshot 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Disclaimer />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": null,
    "nodeType": "host",
    "props": Object {
      "children": <p>
        Data provided by Marvel. © 2014 Marvel
</p>,
      "className": "disclaimer",
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": null,
      "nodeType": "host",
      "props": Object {
        "children": "Data provided by Marvel. © 2014 Marvel",
      },
      "ref": null,
      "rendered": "Data provided by Marvel. © 2014 Marvel",
      "type": "p",
    },
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": null,
      "nodeType": "host",
      "props": Object {
        "children": <p>
          Data provided by Marvel. © 2014 Marvel
</p>,
        "className": "disclaimer",
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": null,
        "nodeType": "host",
        "props": Object {
          "children": "Data provided by Marvel. © 2014 Marvel",
        },
        "ref": null,
        "rendered": "Data provided by Marvel. © 2014 Marvel",
        "type": "p",
      },
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;

exports[`Loading matches its snapshot 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Loading />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": null,
    "nodeType": "host",
    "props": Object {
      "children": "Loading super-heroes...",
    },
    "ref": null,
    "rendered": "Loading super-heroes...",
    "type": "p",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": null,
      "nodeType": "host",
      "props": Object {
        "children": "Loading super-heroes...",
      },
      "ref": null,
      "rendered": "Loading super-heroes...",
      "type": "p",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
